src/Common/getNumberOfPhysicalCPUCores.cpp

```diff
@@ -36,20 +42,73 @@ static uint32_t getCGroupLimitedCPUCores(unsigned default_cpu_count)
 {
-    unsigned cpu_count = std::thread::hardware_concurrency();
+    /// The CPUID instruction isn't reliable across different vendors and CPU models. The best option to get the physical core count is
+    /// to parse /proc/cpuinfo. boost::thread::physical_concurrency() does the same, so use their implementation.
+    ///
+    /// See https://doc.callmematthi.eu/static/webArticles/Understanding%20Linux%20_proc_cpuinfo.pdf
+    std::ifstream proc_cpuinfo("/proc/cpuinfo");
+
+    if (!proc_cpuinfo.is_open())
+        /// In obscure cases (chroot) /proc can be unmounted
+        return std::thread::hardware_concurrency();
+
+    using CoreEntry = std::pair<size_t, size_t>; /// physical id, core id
+    using CoreEntrySet = std::set<CoreEntry>;
+
+    CoreEntrySet core_entries;
+
+    CoreEntry cur_core_entry;
+    std::string line;
+
+    while (std::getline(proc_cpuinfo, line))
+    {
+        size_t pos = line.find(std::string(":"));
+        if (pos == std::string::npos)
+            continue;
+
+        std::string key = line.substr(0, pos);
+        std::string val = line.substr(pos + 1);
+
+        if (key.find("physical id") != std::string::npos)
+        {
+            cur_core_entry.first = std::stoi(val);
+            continue;
+        }
 
-    /// Most of x86_64 CPUs have 2-way Hyper-Threading
+        if (key.find("core id") != std::string::npos)
+        {
+            cur_core_entry.second = std::stoi(val);
+            core_entries.insert(cur_core_entry);
+            continue;
+        }
+    }
+    return core_entries.empty() ? /*unexpected format*/ std::thread::hardware_concurrency() : static_cast<unsigned>(core_entries.size());
+}
+catch (...)
+{
+    return std::thread::hardware_concurrency(); /// parsing error
+}
+#endif
+
+unsigned getNumberOfPhysicalCPUCoresImpl()
+{
+    unsigned cpu_count = std::thread::hardware_concurrency(); /// logical cores (with SMT/HyperThreading)
+
+    /// Most x86_64 CPUs have 2-way SMT (Hyper-Threading).
     /// Aarch64 and RISC-V don't have SMT so far.
-    /// POWER has SMT and it can be multiple way (like 8-way), but we don't know how ClickHouse really behaves, so use all of them.
+    /// POWER has SMT and it can be multi-way (e.g. 8-way), but we don't know how ClickHouse really behaves, so use all of them.
 
-#if defined(__x86_64__)
-    /// Let's limit ourself to the number of physical cores.
-    /// But if the number of logical cores is small - maybe it is a small machine
-    /// or very limited cloud instance and it is reasonable to use all the cores.
+#if defined(__x86_64__) && defined(OS_LINUX)
+    /// On really big machines, SMT is detrimental to performance (+ ~5% overhead in ClickBench). On such machines, we limit ourself to the physical cores.
+    /// Few cores indicate it is a small machine, runs in a VM or is a limited cloud instance --> it is reasonable to use all the cores.
     if (cpu_count >= 32)
-        cpu_count /= 2;
+        cpu_count = physical_concurrency();
```

Comment:
This looks heavy. Let's use a static variable to parse the /proc/cpuinfo once?

Reply:
Good idea, it is implemented already (l. 126).